Source/Game/Game/Editor/AssetBrowser.cpp

```diff
@@ -0,0 +1,490 @@
+
+        ImGui::PushStyleVar(ImGuiStyleVar_FramePadding, ImVec2(0, 0));
+        auto cursorPos = ImGui::GetCursorPos();
+        if (ImGui::Button("##", buttonSize)) // background button
+        {
+            ItemClicked(item);
+        }
+        if (ImGui::IsItemHovered())
+        {
+            ImGui::SetTooltip("%s", itemName.c_str());
+            ItemHovered(item);
+        }
+        auto endCursorPos = ImGui::GetCursorPos();
+        ImGui::PopStyleVar();
+
+        CalculateAspectRatio(size, defaultSize);
+
+        ImVec2 imagePos = ImVec2((buttonSize.x - size.x) * 0.5f + cursorPos.x,
+            ((buttonSize.y - textHeight) - size.y) * 0.5f + cursorPos.y);
+        ImGui::SetCursorPos(imagePos);
+        ImGui::Image(imageTexture, size);
+        ImGui::SetCursorPosX((buttonSize.x - defaultSize.x) * 0.5f + cursorPos.x);
+        ImGui::SetCursorPosY(defaultSize.y + cursorPos.y + ImGui::GetStyle().FramePadding.y * 2.5f * scale);
+        ImGui::Text("%s", ellipsedText.c_str());
+
+        ImGui::EndGroup();
+
+        ImGui::SetCursorPos(endCursorPos);
+        ImGui::SameLine();
+        if (!CanDisplayMore(buttonSize))
+            ImGui::NewLine();
+    }
+
+    void AssetBrowser::ProcessDisplayRendering(const fs::path &item, i32 i)
+    {
+        auto ext = item.extension().string();
+
+        if (IsTexture(ext))
+        {
+            if (!_images[i])
+            {
+                Renderer::TextureDesc textureDesc;
+                textureDesc.path = item.string();
+                auto textureID = _renderer->LoadTexture(textureDesc);
+                _images[i] = _renderer->GetImguiImageHandle(textureID);
+                _imagesSize[i] = ImVec2((float)_renderer->GetTextureWidth(textureID),
+                    (float)_renderer->GetTextureHeight(textureID));
+            }
+
+            _currentImage = _images[i];
+            _currentSize = _imagesSize[i];
+            return;
+        }
+
+        _currentImage = _defaultImageHandle;
+        _currentSize = _defaultDisplaySize;
+    }
+
+    void AssetBrowser::CalculateAspectRatio(ImVec2 &size, const ImVec2 ref)
+    {
+        bool inverse = true;
+        float max = size.y;
+        float min = size.x;
+
+        if (size.x >= size.y)
+        {
+            max = size.x;
+            min = size.y;
+            inverse = false;
+        }
+
+        float ratio = max / ref.x;
+
+        // if the texture is too small,
+        // the resize ratio is blocked at x2.5
+        // to not make it too blurry
+        if (ratio < 0.4f)
+            ratio = 0.4f;
+
+        max /= ratio;
+        min /= ratio;
+
+        if (inverse)
+        {
+            size.x = min;
+            size.y = max;
+        }
+        else
+        {
+            size.x = max;
+            size.y = min;
+        }
+    }
+
+    bool AssetBrowser::SearchString(const std::string &ref, const std::string &key, bool insensitive)
```

Comment:
This looks like something that could go into StringUtil.h, I believe it's in the Engine submodule